File: tests/test_similar.js
const assert = require("node:assert/strict");
const test = require("node:test");

require("../extension/similar.js");

const {
  MAX_RECOMMENDATIONS,
  DEFAULT_VISIBLE_RECOMMENDATIONS,
  SIMILAR_CACHE_VERSION,
  alternativeCandidates,
  confidenceLabel,
  mapsSearchUrl,
  similarMapTargetPayload,
  similarDiagnosticsSummary,
  similarDisplayState,
  similarPayload,
  sortedSimilarRecommendations,
  visibleSimilarRecommendations,
} = globalThis.MeshiLensSimilar;

test("omits the selected Tabelog listing from manual candidate choices", () => {
  const selected = { url: "https://tabelog.com/tw/hyogo/A2803/A280303/28071372/" };
  const candidates = [
    { name: "麥當勞", url: "https://tabelog.com/hyogo/A2803/A280303/28071372/" },
    { name: "另一家店", url: "https://tabelog.com/hyogo/A2803/A280303/28000001/" },
  ];
  assert.deepEqual(alternativeCandidates(candidates, selected), [candidates[1]]);
});

test("does not show low-confidence Tabelog matches as manual choices", () => {
  const candidates = [
    { name: "低信心店", url: "https://tabelog.com/hyogo/A2803/A280303/28000001/", confidence: "low" },
    { name: "待確認店", url: "https://tabelog.com/hyogo/A2803/A280303/28000002/", confidence: "medium" },
  ];
  assert.deepEqual(alternativeCandidates(candidates, null), [candidates[1]]);
});

test("uses confidence labels and Google Maps place searches for UI links", () => {
  assert.equal(confidenceLabel("high"), "高信心");
  assert.equal(confidenceLabel("medium"), "待確認");
  assert.equal(confidenceLabel("low"), "低信心");
  assert.equal(
    mapsSearchUrl("鮨 みなと", "東京都中央区銀座"),
    "https://www.google.com/maps/search/?api=1&query=%E9%AE%A8%20%E3%81%BF%E3%81%AA%E3%81%A8%20%E6%9D%B1%E4%BA%AC%E9%83%BD%E4%B8%AD%E5%A4%AE%E5%8C%BA%E9%8A%80%E5%BA%A7",
  );
  assert.deepEqual(
    similarMapTargetPayload({ name: "MENSHO", url: "https://tabelog.com/tokyo/A1323/A132302/13203848/" }),
    { name: "MENSHO", url: "https://tabelog.com/tokyo/a1323/a132302/13203848/" },
  );
  assert.equal(similarMapTargetPayload({ name: "MENSHO", url: "https://example.com/" }), null);
});

test("builds a bounded similar-restaurant request from selected Tabelog facts", () => {
  assert.equal(MAX_RECOMMENDATIONS, 6);
  assert.equal(DEFAULT_VISIBLE_RECOMMENDATIONS, 3);
  assert.equal(SIMILAR_CACHE_VERSION, "nearby-v14");
  assert.deepEqual(
    similarPayload({
      name: "鮨 みなと",
      url: "https://tabelog.com/tokyo/A1301/A130101/1300001/",
      genres: ["寿司", "日本料理"],
      station: "銀座駅",
      address: "東京都中央区銀座",
      dinner_price: "￥20,000～￥29,999",
    }),
    {
      selected: {
        name: "鮨 みなと",
        url: "https://tabelog.com/tokyo/A1301/A130101/1300001/",
        genres: ["寿司", "日本料理"],
        station: "銀座駅",
        address: "東京都中央区銀座",
        lunch_price: "",
        dinner_price: "￥20,000～￥29,999",
      },
    },
  );
});

test("sorts already-loaded similar restaurants without another request", () => {
  const recommendations = [
    { name: "評論最多拉麵", similarity_score: 92, rating: 3.2, review_count: 500, genres: ["ラーメン"], genre_labels: ["拉麵"] },
    { name: "高分壽司", similarity_score: 71, rating: 3.9, review_count: 12, genres: ["寿司"], genre_labels: ["壽司"] },
    { name: "中分拉麵", similarity_score: 80, rating: 3.6, review_count: 100, genres: ["ラーメン"], genre_labels: ["拉麵"] },
  ];
  assert.deepEqual(
    sortedSimilarRecommendations(recommendations, "rating").map((item) => item.name),
    ["高分壽司", "中分拉麵", "評論最多拉麵"],
  );
  assert.deepEqual(
    visibleSimilarRecommendations(recommendations, { sort: "reviews", expanded: true }).map((item) => item.name),
    ["評論最多拉麵", "中分拉麵", "高分壽司"],
  );
});

test("does not request similar restaurants without Tabelog URL or cuisine", () => {
  assert.equal(similarPayload({ name: "鮨 みなと", genres: ["寿司"] }), null);
  assert.equal(similarPayload({ name: "鮨 みなと", url: "https://tabelog.com/example/" }), null);
});

test("keeps an explicit empty state and safe diagnostics when no nearby recommendation is verified", () => {
  const diagnostics = { search_scope: "銀座駅", returned_count: 5, unverified_location_count: 4, below_quality_count: 1 };
  assert.deepEqual(similarDisplayState([], diagnostics), { status: "empty", diagnostics });
  assert.deepEqual(similarDisplayState(null), { status: "empty", diagnostics: null });
  assert.deepEqual(
    similarDisplayState([{ name: "附近店家" }]),
    { status: "ready", recommendations: [{ name: "附近店家" }], diagnostics: null, sort: "recommended", expanded: false },
  );
  assert.equal(
    similarDiagnosticsSummary(diagnostics),
    "搜尋範圍：銀座駅；Tabelog 回傳 5 家；1 家未達相似度門檻。",
  );
  assert.equal(similarDiagnosticsSummary(null), "");
});

File: tests/test_similar.py
import unittest

from meshi_lens.similar import (
    rank_similar_candidates,
    rank_similar_candidates_with_diagnostics,
)
from meshi_lens.localization import tabelog_genre_zh_hant_map


class SimilarRankingTests(unittest.TestCase):
    def test_cuisine_translation_map_is_loaded_from_editable_data_file(self) -> None:
        self.assertEqual(tabelog_genre_zh_hant_map()["アメリカ料理"], "美式料理")

    def test_prefers_same_cuisine_station_and_budget_without_fetching_details(self) -> None:
        seed = {
            "url": "https://tabelog.com/tokyo/A1301/A130101/1300001/",
            "genres": ["寿司"],
            "station": "銀座駅",
            "address": "東京都中央区銀座",
            "dinner_price": "￥20,000～￥29,999",
        }
        ranked = rank_similar_candidates(
            seed,
            [
                {"name": "元の店", "url": seed["url"], "genres": ["寿司"]},
                {
                    "name": "銀座の寿司店",
                    "url": "https://tabelog.com/tokyo/A1301/A130101/1300002/",
                    "genres": ["寿司"],
                    "station": "銀座駅",
                    "address": "東京都中央区銀座1-1-1",
                    "rating": 3.81,
                    "review_count": 210,
                    "dinner_price": "￥20,000～￥29,999",
                },
                {
                    "name": "別の寿司店",
                    "url": "https://tabelog.com/tokyo/A1301/A130101/1300003/",
                    "genres": ["寿司"],
                    "station": "新橋駅",
                    "rating": 3.95,
                    "review_count": 800,
                    "dinner_price": "￥50,000～￥59,999",
                },
            ],
        )
        self.assertEqual([item["name"] for item in ranked], ["銀座の寿司店"])
        self.assertEqual(ranked[0]["reasons"], ["同為壽司", "同為銀座站", "晚餐價位接近"])
        self.assertEqual(ranked[0]["address"], "東京都中央区銀座1-1-1")
        self.assertNotIn("元の店", [item["name"] for item in ranked])

    def test_rejects_a_same_cuisine_restaurant_in_a_different_prefecture(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1301/A130101/1300001/",
                "genres": ["寿司"],
                "station": "銀座駅",
                "address": "東京都中央区銀座1-1-1",
            },
            [
                {
                    "name": "札幌の寿司店",
                    "url": "https://tabelog.com/hokkaido/A0101/A010101/1000001/",
                    "genres": ["寿司"],
                    "station": "札幌駅",
                    "address": "北海道札幌市中央区1-1",
                    "rating": 4.2,
                    "review_count": 5000,
                }
            ],
        )
        self.assertEqual(ranked, [])

    def test_accepts_a_nearby_different_station_in_the_same_ward(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1301/A130101/1300001/",
                "genres": ["寿司"],
                "station": "銀座駅",
                "address": "東京都中央区銀座1-1-1",
            },
            [
                {
                    "name": "築地の寿司店",
                    "url": "https://tabelog.com/tokyo/A1301/A130101/1300002/",
                    "genres": ["寿司"],
                    "station": "築地駅",
                    "address": "東京都中央区築地2-2-2",
                    "rating": 3.8,
                    "review_count": 100,
                }
            ],
        )
        self.assertEqual(ranked[0]["reasons"][:2], ["同為壽司", "同一行政區"])

    def test_matches_a_maps_line_qualified_station_to_tabelog_station_name(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1305/A130501/1300001/",
                "genres": ["ハンバーガー"],
                "station": 'JR山手線「池袋」駅',
                "address": "東京都豊島区東池袋2-57-2",
            },
            [
                {
                    "name": "池袋のハンバーガー店",
                    "url": "https://tabelog.com/tokyo/A1305/A130501/1300002/",
                    "genres": ["ハンバーガー"],
                    "station": "池袋駅",
                    "rating": 3.6,
                    "review_count": 40,
                }
            ],
        )
        self.assertEqual([item["name"] for item in ranked], ["池袋のハンバーガー店"])
        self.assertEqual(ranked[0]["reasons"][:2], ["同為漢堡", "同為池袋站"])

    def test_matches_a_maps_line_qualified_station_without_quotes(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1323/A132302/1300001/",
                "genres": ["ラーメン"],
                "station": "JR山手線大塚駅",
                "address": "東京都豊島区北大塚1-1-1",
            },
            [
                {
                    "name": "大塚のラーメン店",
                    "url": "https://tabelog.com/tokyo/A1323/A132302/1300002/",
                    "genres": ["ラーメン"],
                    "station": "大塚駅",
                    "rating": 3.6,
                    "review_count": 40,
                }
            ],
        )
        self.assertEqual([item["name"] for item in ranked], ["大塚のラーメン店"])
        self.assertEqual(ranked[0]["reasons"][:2], ["同為拉麵", "同為大塚站"])

    def test_uses_precise_tabelog_area_scope_when_result_card_omits_location(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1323/A132302/13276342/",
                "genres": ["おにぎり"],
                "station": "JR山手線大塚駅",
                "address": "東京都豊島区北大塚2-27-5",
            },
            [
                {
                    "name": "大塚のおにぎり店",
                    "url": "https://tabelog.com/tokyo/A1323/A132302/1300002/",
                    "genres": ["おにぎり"],
                    "rating": 3.6,
                    "review_count": 40,
                }
            ],
            search_area_path="tokyo/A1323/A132302",
        )
        self.assertEqual([item["name"] for item in ranked], ["大塚のおにぎり店"])
        self.assertEqual(
            ranked[0]["reasons"][:2], ["同為おにぎり", "同一 Tabelog 區域"]
        )

    def test_filters_a_nearby_card_without_a_similarity_signal(self) -> None:
        ranked = rank_similar_candidates(
            {"url": "https://tabelog.com/tokyo/A1323/A132302/13276342/", "genres": ["おにぎり"]},
            [{"name": "附近餐廳", "url": "https://tabelog.com/tokyo/A1323/A132302/1300002/", "genres": ["定食"], "rating": 3.5, "review_count": 100}],
            search_area_path="tokyo/A1323/A132302",
        )
        self.assertEqual(ranked, [])

    def test_filters_candidates_without_a_similarity_signal(self) -> None:
        ranked = rank_similar_candidates(
            {"url": "https://tabelog.com/tokyo/A1301/A130101/1300001/", "genres": ["寿司"]},
            [
                {
                    "name": "遠方のカフェ",
                    "url": "https://tabelog.com/tokyo/A1301/A130101/1300002/",
                    "genres": ["カフェ"],
                    "rating": 3.9,
                    "review_count": 500,
                }
            ],
        )
        self.assertEqual(ranked, [])

    def test_reports_aggregate_reasons_when_nearby_filter_excludes_results(self) -> None:
        ranked, diagnostics = rank_similar_candidates_with_diagnostics(
            {
                "url": "https://tabelog.com/tokyo/A1301/A130101/1300001/",
                "genres": ["寿司"],
                "station": "銀座駅",
                "address": "東京都中央区銀座1-1-1",
            },
            [
                {
                    "name": "札幌の寿司店",
                    "url": "https://tabelog.com/hokkaido/A0101/A010101/1000001/",
                    "genres": ["寿司"],
                    "station": "札幌駅",
                    "address": "北海道札幌市中央区1-1",
                }
            ],
        )
        self.assertEqual(ranked, [])
        self.assertEqual(diagnostics["returned_count"], 1)
        self.assertEqual(diagnostics["unverified_location_count"], 1)
        self.assertEqual(diagnostics["recommendation_count"], 0)

    def test_localizes_tabelog_cuisine_in_display_reasons(self) -> None:
        ranked = rank_similar_candidates(
            {
                "url": "https://tabelog.com/tokyo/A1301/A130101/1300001/",
                "genres": ["アメリカ料理"],
                "station": "銀座駅",
            },
            [
                {
                    "name": "美式餐廳",
                    "url": "https://tabelog.com/tokyo/A1301/A130101/1300002/",
                    "genres": ["アメリカ料理"],
                    "station": "銀座駅",
                    "rating": 3.5,
                    "review_count": 20,
                }
            ],
        )
        self.assertEqual(ranked[0]["reasons"][:2], ["同為美式料理", "同為銀座站"])
